source/ui/x11/X11Window.ooc

```diff
@@ -52,8 +64,33 @@ X11Window: class extends NativeWindow {
 	}
 	resize: func (size: IntSize2D) { XResizeWindow(this display, this backend, size width, size height) }
 	free: override func {
+		if (this _xImage)
+			XDestroyImage(this _xImage)
+		XFreeGC(this display, this _x11GraphicsContext)
+		XDestroyWindow(this display, this backend)
 		XCloseDisplay(this display)
 		super()
 	}
+	draw: func (image: RasterYuv420Semiplanar) {
+		if (this _cacheSize != image size) {
+			if (this _xImage)
+				XDestroyImage(this _xImage)
+			imageData := gc_malloc(image width * image height * 4)
+			this _xImage = XCreateImage(this display, CopyFromParent, this _displayDepth, ZPixmap, 0, imageData, image width, image height, 8, 0)
+			this _cacheSize = image size
+		}
+		for (row in 0 .. image height )
+			for (column in 0 .. image width) {
+				pixel := 0 as ULong
+				pixelData := pixel& as UInt8*
+				color := image[column, row] toBgr()
+				pixelData@ = color blue
+				(pixelData + 1)@ = color green
+				(pixelData + 2)@ = color red
+				XPutPixel(this _xImage, column, row, pixel)
```

Comment:
It would clean up the code here a lot. You wouldn't actually allocate the memory each time, since the underlying `ByteBuffer` is recycled when the temporary bgr image is freed.